l10n_fi_reconcile/models/account_bank_statement.py

```diff
@@ -0,0 +1,105 @@
+        precision = st_line_currency and st_line_currency.decimal_places or company_currency.decimal_places
+        field = currency and 'amount_residual_currency' or 'amount_residual'
+        liquidity_field = currency and 'amount_currency' or amount > 0 and 'debit' or 'credit'
+        liquidity_amt_clause = currency and '%(amount)s::numeric' or 'abs(%(amount)s::numeric)'
+        params = {'company_id': self.env.user.company_id.id,
+                  'account_payable_receivable': (
+                    self.journal_id.default_credit_account_id.id,
+                    self.journal_id.default_debit_account_id.id),
+                  'amount': float_round(amount, precision_digits=precision),
+                  'partner_id': self.partner_id.id,
+                  'excluded_ids': tuple(excluded_ids),
+                  'ref': self.ref,
+                  }
+
+        if self.ref:
+            # Check for both matching reference and amount
+            add_to_select = ", CASE WHEN LTRIM(aml.payment_reference, '0') = %(ref)s \
+                            THEN 1 ELSE 2 END as temp_field_order "
+            add_to_from = " JOIN account_move m ON m.id = aml.move_id "
+            select_clause, from_clause, where_clause = self._get_common_sql_query(overlook_partner=True, excluded_ids=excluded_ids, split=True)
+            sql_query = select_clause + add_to_select + from_clause + add_to_from + where_clause
+            add_to_where = " AND (LTRIM(aml.payment_reference, '0') = LTRIM(%(ref)s, '0') or m.name = %(ref)s) \
+                            AND (" + field + " = %(amount)s::numeric OR (acc.internal_type = 'liquidity' \
+                            AND " + liquidity_field + " = " + liquidity_amt_clause + ")) \
+                            ORDER BY temp_field_order, date_maturity asc, aml.id asc"
+
+            self.env.cr.execute(sql_query + add_to_where, params)
+            results = self.env.cr.fetchone()
+            if results:
+                return self.env['account.move.line'].browse(results[0])
+
+            # Check for just a matching reference
+            add_to_where = " AND (LTRIM(aml.payment_reference, '0') = LTRIM(%(ref)s, '0') or m.name = %(ref)s) \
+                            ORDER BY temp_field_order, date_maturity asc, aml.id asc"
+            self.env.cr.execute(sql_query + add_to_where, params)
+            results = self.env.cr.fetchone()
+            if results:
+                return self.env['account.move.line'].browse(results[0])
+
+        # Return super if our queries yielded nothing
+        return super(AccountBankStatementLine, self).get_reconciliation_proposition(excluded_ids=excluded_ids)
+
+    def get_statement_line_for_reconciliation_widget(self):
+        res = super(AccountBankStatementLine, self).get_statement_line_for_reconciliation_widget()
+
+        # partner_name key is already in use so use something else
+        if self.partner_name:
+            res['partner_note'] = self.partner_name
```

Comment:
We can add a new field but we would never make any use of that new field apart from assigning it a new value.
The Qweb doesn't access the account.bank.statement's fields, it only uses the dict we return it. The code in the Qweb is also misleading, it took me a while to realize it doesn't actually use any existing fields, it just uses the dict that this method returns. The problem is we have two different partner names we want to return for display. One of them is an actual `res.partner`'s `name` field, in our case the `name` of a `res.partner` that has a matching `payment_reference`. The second one is the `self.partner_name` which because of the way the bank statement importer uses the field, that might contain pretty much anything. The Qweb already uses the partner_name, so we can't have them both as `partner_name`.

Also the `partner_note` is only referred to here when it is created, and on one line of the Qweb. The Qweb does look like it's using actual fields on a record, but changing the name of this key doesn't remove that problem. The original code is already very confusing https://github.com/odoo/odoo/blob/10.0/addons/account/static/src/xml/account_reconciliation.xml#L235